File: geometry/MatchMaking.java
import java.util.*;

public class MatchMaking {

	public String makeMatch(String[] namesWomen, String[] answersWomen, String[] namesMen,
						    String[] answersMen, String queryWomen) {
		int n = namesWomen.length;
		boolean[] validWomen = new boolean[n];
		boolean[] validMen = new boolean[n];
		Arrays.fill(validWomen, true);
		Arrays.fill(validMen, true);
		while (true) {
			int selWomen = -1;
			for (int i = 0; i < n; ++i) {
				if (validWomen[i] && (selWomen == -1
					|| namesWomen[i].compareTo(namesWomen[selWomen]) < 0)) {
					selWomen = i;
				}
			}
			int selMen = -1;
			int bestMatch = -1;
			for (int i = 0; i < n; ++i) {
				if (!validMen[i]) {
					continue;
				}
				int match = getMatch(answersWomen[selWomen], answersMen[i]);
				if (match > bestMatch || (match == bestMatch &&
					namesMen[i].compareTo(namesMen[selMen]) < 0)) {
					selMen = i;
					bestMatch = match;
				}
			}
			if (namesWomen[selWomen].equals(queryWomen)) {
				return namesMen[selMen];
			}
			validWomen[selWomen] = false;
			validMen[selMen] = false;
		}
	}
	
	int getMatch(String a, String b) {
		int result = 0;
		for (int i = 0; i < a.length(); ++i) {
			if (a.charAt(i) == b.charAt(i)) {
				result++;
			}
		}
		return result;
	}
}

File: tutorials/MatchMaking.cpp
// SRM 203 Div 1 Easy
#include <algorithm>
#include <string>
#include <vector>

using namespace std;

class MatchMaking {
public:
	string makeMatch(vector<string> namesWomen, vector<string> answersWomen,
					 vector<string> namesMen, vector<string> answersMen,
					 string queryWoman) {
		size_t size = namesWomen.size();
		vector<bool> usedWomen(size, false);
		vector<bool> usedMen(size, false);
		for (size_t used = 0; used < size; ++used) {
			size_t selWoman = -1;
			for (size_t i = 0; i < size; ++i) {
				if (!usedWomen[i] && (selWoman == -1 ||
					namesWomen[i] < namesWomen[selWoman])) {
					selWoman = i;
				}
			}
			size_t selMan = -1;
			size_t match = 0;
			for (size_t i = 0; i < size; ++i) {
				if (!usedMen[i]) {
					size_t m = getMatch(answersWomen[selWoman], answersMen[i]);
					if (selMan == -1 || match < m || (match == m &&
						namesMen[i] < namesMen[selMan])) {
						selMan = i;
						match = m;
					}
				}
			}
			usedWomen[selWoman] = true;
			usedMen[selMan] = true;
			if (namesWomen[selWoman] == queryWoman) {
				return namesMen[selMan];
			}
		}
		return "";
	}
	
	size_t getMatch(string a, string b) {
		size_t m = 0;
		for (size_t i = 0; i < a.size(); ++i) {
			if (a[i] == b[i]) {
				m++;
			}
		}
		return m;
	}
};
